src/server/middlewares/validateVote.js

```diff
@@ -1,175 +1,260 @@
   }
-};
+}
 
-const checkFestival = async (vote) => {
-  const valid = await isActiveFestival(vote.festival);
-  if (!valid) {
-    throw Error('invalid festival');
+async function checkFestival(vote, questionContract) {
+  const festivalChainId = await questionContract.methods.festival().call();
+  const isValid = await isActiveFestival(festivalChainId);
+
+  if (!isValid) {
+    throw Error('Invalid festival');
   }
-};
+}
 
-const checkNonce = async (vote) => {
-  const isValidNonce = await admin.methods
-    .isValidVotingNonce(vote.booth, vote.nonce)
+async function checkNonce(vote) {
+  const isValidNonce = await adminContract.methods
+    .isValidVotingNonce(vote.boothAddress, vote.nonce)
     .call();
+
   if (!isValidNonce) {
-    throw Error('invalid nonce');
+    throw Error('Invalid nonce');
   }
-};
+}
 
-const checkSigs = async (vote) => {
+async function checkSignatures(vote) {
   if (
     !isSignatureValid(
       packVote(
-        vote.festivalAnswers.map((a) => a.id),
+        vote.festivalAnswerIds,
         vote.festivalVoteTokens,
-        vote.artworkAnswers.map((a) => a.id),
+        vote.artworkAnswerIds,
         vote.artworkVoteTokens,
       ),
-      vote.signature,
-      vote.sender,
-    ) ||
+      vote.senderSignature,
+      vote.senderAddress,
+    )
+  ) {
+    throw Error('Invalid sender signature');
+  }
+
+  if (
     !isSignatureValid(
-      packBooth(
-        vote.festivalAnswers.map((a) => a.id),
-        vote.nonce,
-      ),
+      packBooth(vote.festivalAnswerIds, vote.nonce),
       vote.boothSignature,
-      vote.booth,
+      vote.boothAddress,
     )
   ) {
-    throw Error('invalid sig');
+    throw Error('Invalid booth signature');
   }
-};
+}
 
-const checkMaxVotes = async (answers, votes, question) => {
+async function checkMaxVotes(
+  answerIds,
+  answerChainIds,
+  voteTokens,
+  questionContract,
+) {
+  const maxVoteTokens = await questionContract.methods.maxVoteTokens().call();
   const seen = {};
-  const maxVoteTokens = await question.methods.maxVoteTokens().call();
   let sum = 0;
+
   await Promise.all(
-    answers.map(async (answer) => {
-      if (seen[answer.id]) {
-        throw Error('repeated answer');
+    answerIds.map(async (answerId, index) => {
+      if (answerId in seen) {
+        throw Error('Duplicate answer');
       }
-      seen[answer.id] = true;
-      sum += votes;
 
-      const a = await question.methods.getAnswer(answer.chainId).call();
-      if (a[0] !== true) {
-        throw Error('invalid answer');
+      seen[answerId] = true;
+      sum += voteTokens[index];
+
+      const result = await questionContract.methods
+        .getAnswer(answerChainIds[index])
+        .call();
+      const isInitialized = result[0];
+
+      if (!isInitialized) {
+        throw Error('Invalid answer');
       }
     }),
   );
-  if (sum > maxVoteTokens * answers.length) {
-    throw Error('too many votes');
+
+  if (sum > maxVoteTokens) {
```

Comment:
Max vote tokens is per answer, users get a flexible amount of vote tokens based on how many artworks they saw